File: static/swift/AirtableQuery.swift
import Foundation

/// Query parameters for Airtable list-records requests. Builder-style:
/// create, chain mutations, pass to a table's list method.
///
/// `Sendable` so instances can cross actor boundaries (e.g., from a model
/// main-actor call into the `AirtableClient` actor).
public struct AirtableQuery: Sendable, Equatable {
    public var formula: String?
    public var sort: [Sort]
    public var fields: [String]
    public var maxRecords: Int?
    public var pageSize: Int?
    public var view: String?
    public var returnFieldsByFieldId: Bool
    public var cellFormat: CellFormat
    public var timeZone: String?
    public var userLocale: String?

    public enum CellFormat: String, Sendable {
        case json
        case string
    }

    public init(
        formula: String? = nil,
        sort: [Sort] = [],
        fields: [String] = [],
        maxRecords: Int? = nil,
        pageSize: Int? = nil,
        view: String? = nil,
        returnFieldsByFieldId: Bool = true,
        cellFormat: CellFormat = .json,
        timeZone: String? = nil,
        userLocale: String? = nil
    ) {
        self.formula = formula
        self.sort = sort
        self.fields = fields
        self.maxRecords = maxRecords
        self.pageSize = pageSize
        self.view = view
        self.returnFieldsByFieldId = returnFieldsByFieldId
        self.cellFormat = cellFormat
        self.timeZone = timeZone
        self.userLocale = userLocale
    }

    // MARK: - Fluent builders

    public func withFormula(_ formula: String?) -> AirtableQuery {
        var copy = self
        copy.formula = formula
        return copy
    }

    public func withSort(_ sort: [Sort]) -> AirtableQuery {
        var copy = self
        copy.sort = sort
        return copy
    }

    public func withSort(field: String, direction: SortDirection = .asc) -> AirtableQuery {
        var copy = self
        copy.sort.append(Sort(field: field, direction: direction))
        return copy
    }

    public func withFields(_ fields: [String]) -> AirtableQuery {
        var copy = self
        copy.fields = fields
        return copy
    }

    public func withMaxRecords(_ maxRecords: Int?) -> AirtableQuery {
        var copy = self
        copy.maxRecords = maxRecords
        return copy
    }

    public func withPageSize(_ pageSize: Int?) -> AirtableQuery {
        var copy = self
        copy.pageSize = pageSize
        return copy
    }

    public func withView(_ view: String?) -> AirtableQuery {
        var copy = self
        copy.view = view
        return copy
    }

    // MARK: - URL encoding

    /// Convert to `[URLQueryItem]` using Airtable's list-records parameter shape:
    /// `filterByFormula`, `fields[]`, `sort[i][field]`, `sort[i][direction]`,
    /// `maxRecords`, `pageSize`, `view`, `returnFieldsByFieldId`, `cellFormat`,
    /// `timeZone`, `userLocale`.
    public func toQueryItems() -> [URLQueryItem] {
        var items: [URLQueryItem] = []

        if let formula = formula, !formula.isEmpty {
            items.append(URLQueryItem(name: "filterByFormula", value: formula))
        }

        for field in fields {
            items.append(URLQueryItem(name: "fields[]", value: field))
        }

        for (i, s) in sort.enumerated() {
            items.append(URLQueryItem(name: "sort[\(i)][field]", value: s.field))
            items.append(URLQueryItem(name: "sort[\(i)][direction]", value: s.direction.rawValue))
        }

        if let maxRecords = maxRecords {
            items.append(URLQueryItem(name: "maxRecords", value: String(maxRecords)))
        }

        if let pageSize = pageSize {
            items.append(URLQueryItem(name: "pageSize", value: String(pageSize)))
        }

        if let view = view, !view.isEmpty {
            items.append(URLQueryItem(name: "view", value: view))
        }

        if returnFieldsByFieldId {
            items.append(URLQueryItem(name: "returnFieldsByFieldId", value: "true"))
        }

        if cellFormat != .json {
            items.append(URLQueryItem(name: "cellFormat", value: cellFormat.rawValue))
        }

        if let tz = timeZone, !tz.isEmpty {
            items.append(URLQueryItem(name: "timeZone", value: tz))
        }

        if let locale = userLocale, !locale.isEmpty {
            items.append(URLQueryItem(name: "userLocale", value: locale))
        }

        return items
    }
}

File: static/java/AirtableQuery.java
// ==========================================
// MyAirtable static runtime.
// ==========================================

package myairtable;

import java.util.ArrayList;
import java.util.List;

/**
 * Query parameters for Airtable list-records requests.
 *
 * <p>Immutable; idiomatic Java construction is the fluent wither chain:
 *
 * <pre>{@code
 * new AirtableQuery()
 *     .withFormula("NOT({Name}=BLANK())")
 *     .withMaxRecords(10)
 *     .withSort("Name", SortDirection.DESC)
 * }</pre>
 */
public final class AirtableQuery {

  /** Cell format for list queries. */
  public enum CellFormat {
    JSON("json"),
    STRING("string");

    private final String wire;

    CellFormat(String wire) {
      this.wire = wire;
    }

    public String wire() {
      return wire;
    }
  }

  /** A single query parameter; keys can repeat ({@code fields[]}, {@code sort[i][...]}). */
  public record Param(String name, String value) {}

  private final String formula;
  private final List<Sort> sort;
  private final List<String> fields;
  private final Integer maxRecords;
  private final Integer pageSize;
  private final String view;
  private final boolean returnFieldsByFieldId;
  private final CellFormat cellFormat;
  private final String timeZone;
  private final String userLocale;

  /** The default query: no filter, ID-keyed fields, JSON cell format. */
  public AirtableQuery() {
    this(null, List.of(), List.of(), null, null, null, true, CellFormat.JSON, null, null);
  }

  private AirtableQuery(
      String formula,
      List<Sort> sort,
      List<String> fields,
      Integer maxRecords,
      Integer pageSize,
      String view,
      boolean returnFieldsByFieldId,
      CellFormat cellFormat,
      String timeZone,
      String userLocale) {
    this.formula = formula;
    this.sort = List.copyOf(sort);
    this.fields = List.copyOf(fields);
    this.maxRecords = maxRecords;
    this.pageSize = pageSize;
    this.view = view;
    this.returnFieldsByFieldId = returnFieldsByFieldId;
    this.cellFormat = cellFormat;
    this.timeZone = timeZone;
    this.userLocale = userLocale;
  }

  // ---- getters ----

  public String getFormula() {
    return formula;
  }

  public List<Sort> getSort() {
    return sort;
  }

  public List<String> getFields() {
    return fields;
  }

  public Integer getMaxRecords() {
    return maxRecords;
  }

  public Integer getPageSize() {
    return pageSize;
  }

  public String getView() {
    return view;
  }

  public boolean isReturnFieldsByFieldId() {
    return returnFieldsByFieldId;
  }

  public CellFormat getCellFormat() {
    return cellFormat;
  }

  public String getTimeZone() {
    return timeZone;
  }

  public String getUserLocale() {
    return userLocale;
  }

  // ---- withers ----

  public AirtableQuery withFormula(String formula) {
    return new AirtableQuery(
        formula,
        sort,
        fields,
        maxRecords,
        pageSize,
        view,
        returnFieldsByFieldId,
        cellFormat,
        timeZone,
        userLocale);
  }

  /** Append a sort clause (ascending). */
  public AirtableQuery withSort(String field) {
    return withSort(field, SortDirection.ASC);
  }

  /** Append a sort clause. */
  public AirtableQuery withSort(String field, SortDirection direction) {
    List<Sort> next = new ArrayList<>(sort);
    next.add(new Sort(field, direction));
    return new AirtableQuery(
        formula,
        next,
        fields,
        maxRecords,
        pageSize,
        view,
        returnFieldsByFieldId,
        cellFormat,
        timeZone,
        userLocale);
  }

  /** Restrict the response to the given field IDs/names. */
  public AirtableQuery withFields(List<String> fields) {
    return new AirtableQuery(
        formula,
        sort,
        fields,
        maxRecords,
        pageSize,
        view,
        returnFieldsByFieldId,
        cellFormat,
        timeZone,
        userLocale);
  }

  public AirtableQuery withMaxRecords(Integer maxRecords) {
    return new AirtableQuery(
        formula,
        sort,
        fields,
        maxRecords,
        pageSize,
        view,
        returnFieldsByFieldId,
        cellFormat,
        timeZone,
        userLocale);
  }

  public AirtableQuery withPageSize(Integer pageSize) {
    return new AirtableQuery(
        formula,
        sort,
        fields,
        maxRecords,
        pageSize,
        view,
        returnFieldsByFieldId,
        cellFormat,
        timeZone,
        userLocale);
  }

  /** Scope the query to a view by raw view ID or name. */
  public AirtableQuery withView(String view) {
    return new AirtableQuery(
        formula,
        sort,
        fields,
        maxRecords,
        pageSize,
        view,
        returnFieldsByFieldId,
        cellFormat,
        timeZone,
        userLocale);
  }

  /**
   * Scope the query to a generated view constant: {@code new
   * AirtableQuery().withView(ContactsView.ACTIVE)} instead of reaching for {@code .getId()}.
   */
  public AirtableQuery withView(AirtableView view) {
    return withView(view.getId());
  }

  public AirtableQuery withReturnFieldsByFieldId(boolean returnFieldsByFieldId) {
    return new AirtableQuery(
        formula,
        sort,
        fields,
        maxRecords,
        pageSize,
        view,
        returnFieldsByFieldId,
        cellFormat,
        timeZone,
        userLocale);
  }

  public AirtableQuery withCellFormat(CellFormat cellFormat) {
    return new AirtableQuery(
        formula,
        sort,
        fields,
        maxRecords,
        pageSize,
        view,
        returnFieldsByFieldId,
        cellFormat,
        timeZone,
        userLocale);
  }

  public AirtableQuery withTimeZone(String timeZone) {
    return new AirtableQuery(
        formula,
        sort,
        fields,
        maxRecords,
        pageSize,
        view,
        returnFieldsByFieldId,
        cellFormat,
        timeZone,
        userLocale);
  }

  public AirtableQuery withUserLocale(String userLocale) {
    return new AirtableQuery(
        formula,
        sort,
        fields,
        maxRecords,
        pageSize,
        view,
        returnFieldsByFieldId,
        cellFormat,
        timeZone,
        userLocale);
  }

  /**
   * Convert to query parameters using Airtable's list-records shape: {@code filterByFormula},
   * {@code fields[]}, {@code sort[i][field]}, {@code sort[i][direction]}, {@code maxRecords},
   * {@code pageSize}, {@code view}, {@code returnFieldsByFieldId}, {@code cellFormat}, {@code
   * timeZone}, {@code userLocale}.
   */
  public List<Param> toParameters() {
    List<Param> params = new ArrayList<>();
    if (formula != null && !formula.isEmpty()) {
      params.add(new Param("filterByFormula", formula));
    }
    for (String field : fields) {
      params.add(new Param("fields[]", field));
    }
    for (int i = 0; i < sort.size(); i++) {
      params.add(new Param("sort[" + i + "][field]", sort.get(i).field()));
      params.add(new Param("sort[" + i + "][direction]", sort.get(i).direction().wire()));
    }
    if (maxRecords != null) {
      params.add(new Param("maxRecords", maxRecords.toString()));
    }
    if (pageSize != null) {
      params.add(new Param("pageSize", pageSize.toString()));
    }
    if (view != null && !view.isEmpty()) {
      params.add(new Param("view", view));
    }
    if (returnFieldsByFieldId) {
      params.add(new Param("returnFieldsByFieldId", "true"));
    }
    if (cellFormat != CellFormat.JSON) {
      params.add(new Param("cellFormat", cellFormat.wire()));
    }
    if (timeZone != null && !timeZone.isEmpty()) {
      params.add(new Param("timeZone", timeZone));
    }
    if (userLocale != null && !userLocale.isEmpty()) {
      params.add(new Param("userLocale", userLocale));
    }
    return params;
  }
}
